util/collections/machine_filters.go

```diff
@@ -214,3 +216,30 @@ func MatchesKubernetesVersion(kubernetesVersion string) Func {
 		return *machine.Spec.Version == kubernetesVersion
 	}
 }
+
+// ValidVersion returns a filter to find machine that have a non empty and valid version.
+func ValidVersion() Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		if machine.Spec.Version == nil {
+			return false
+		}
+		if _, err := semver.ParseTolerant(*machine.Spec.Version); err != nil {
+			return false
+		}
+		return true
+	}
```

Comment:
The `spec` field might also contain some other information that we don't need, for example if a vendor is running their own versioning scheme based on Kubernetes (e.g. `1.21.1-xyz.0`), we should make sure to output 1.21.1.

Other than that, instead of using the Machine's spec field, maybe we can have the machine controller inspect the Node object and return the version that's available under `Node.Status.NodeInfo.KubeletVersion`?

Reply:
We could also consider exposing the entire NodeSystemInfo struct in Machine.Status. @CecileRobertMichon @fabriziopandini thoughts?

Reply:
> The spec field might also contain some other information that we don't need, for example if a vendor is running their own versioning scheme based on Kubernetes (e.g. 1.21.1-xyz.0), we should make sure to output 1.21.1.

If we are removing xyz.0, then we won't detect properly 1.21.1-xyz.0 --> 1.21.1-xyz.1, which is something I expect vendors expects to do.

if we are considering 1.21.1-xyz.0 from a semver PoV it is lower than 1.21.1; however this should not be a problem, because I expect that once one vendor adopts this versioning schema, it will do it consistently 

> Other than that, instead of using the Machine's spec field, maybe we can have the machine controller inspect the Node object and return the version that's available under Node.Status.NodeInfo.KubeletVersion?

I don't see this gives better guarantees, because while MachineVersion is used by capbk/kubeadm to create the control plane of a given version, KubeletVersion instead is a consequence of what is installed in the images and it has no direct connections with the api server version.



> We could also consider exposing the entire NodeSystemInfo struct in Machine.Status. @CecileRobertMichon @fabriziopandini thoughts?

I'm +1 for this change; if there is agreement I will open an issue so we can tackle this in a separated version.

Reply:
> The spec field might also contain some other information that we don't need, for example if a vendor is running their own versioning scheme based on Kubernetes (e.g. 1.21.1-xyz.0), we should make sure to output 1.21.1.

If we are removing xyz.0, then we won't detect properly 1.21.1-xyz.0 --> 1.21.1-xyz.1, which is something I expect vendors expects to do.

if we are considering 1.21.1-xyz.0 from a semver PoV it is lower than 1.21.1; however this should not be a problem, because I expect that once one vendor adopts this versioning schema, it will do it consistently 

> Other than that, instead of using the Machine's spec field, maybe we can have the machine controller inspect the Node object and return the version that's available under Node.Status.NodeInfo.KubeletVersion?

I don't see this gives better guarantees, because while MachineVersion is used by capbk/kubeadm to create the control plane of a given version, KubeletVersion instead is a consequence of what is installed in the images and it has no direct connections with the api server version.



> We could also consider exposing the entire NodeSystemInfo struct in Machine.Status. @CecileRobertMichon @fabriziopandini thoughts?

I'm +1 for this change; if there is agreement I will open an issue so we can tackle this in a separated version.

Reply:
>If we are removing xyz.0, then we won't detect properly 1.21.1-xyz.0 --> 1.21.1-xyz.1, which is something I expect vendors would like to do.

> if we are considering 1.21.1-xyz.0 from a semver PoV it is lower than 1.21.1; however this should not be a problem, because I expect that once one vendor adopts this versioning schema, it will do it consistently

There are some variants like 1.21.1+xyz.0 which are valid semver but don't seem to not work the way we expect. According to the semver library we are using 1.21.1+xyz.0 and 1.21.1+xyz.1 are considered equal. The 0 or 1 in this case is considered as the build identifier. How about we add a check to also sort by build identifier if versions are equal? (Currently we use Name as the tie breaker).

> if there is agreement I will open an issue so we can tackle this in a separated PR

+1 for separate PR.

Reply:
FYI, PR for adding NodeSystemInfo to Machine - https://github.com/kubernetes-sigs/cluster-api/pull/5079